DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java

```diff
@@ -192,11 +214,12 @@ public static DocumentsRunner.CompletionStatus run(Function<Path,LuceneDocuments
                                                        List<String> indexAllowlist,
                                                        ShardMetadata.Factory shardMetadataFactory,
                                                        SnapshotShardUnpacker.Factory unpackerFactory,
-                                                       long maxShardSizeBytes)
+                                                       long maxShardSizeBytes,
+                                                       RootDocumentMigrationContext rootDocumentContext)
             throws IOException, InterruptedException, NoWorkLeftException {
         var scopedWorkCoordinator = new ScopedWorkCoordinator(workCoordinator, leaseExpireTrigger);
-        confirmShardPrepIsComplete(indexMetadataFactory, snapshotName, indexAllowlist, scopedWorkCoordinator);
-        if (!workCoordinator.workItemsArePending()) {
+        confirmShardPrepIsComplete(indexMetadataFactory, snapshotName, indexAllowlist, scopedWorkCoordinator, rootDocumentContext);
+        if (!workCoordinator.workItemsArePending(rootDocumentContext.getWorkCoordinationContext()::createItemsPendingContext)) {
```

Comment:
This is hacky, we've got a instance that should accept a context and internally it can create the work item pending context.

Reply:
Why is it hacky?  The lifecycle of the context should be controlled by the work coordinator.  It seems more hacky to make every call to the work coordinator be responsible for creating and tearing down contexts around singly-scoped calls.

Reply:
Its hacky because it requires the caller RfsMigrateDocuments needs to know how to generate the correct kind of contexts for the work coordinater, it shouldn't need to know that.  Instead the work coordinator should be context aware - as in passed via construction, then `workItemsArePending` is called creates the appropriate context.

If we were passing a concrete context in function parameters, then I wouldn't have as much of an issue - but in this case a factory is being passed in.  While this is clever so a factory doesn't need to be provided into instance creation it makes this code very hard to interpret.

Ideally we would have code that follows a _single_ pattern of access to the context such as the following.  
```java
var docIndexer = new DocumentIndexer(context);
docIndexer.doIndexing();

// OR but less preferred

var docIndexer = new DocumentIndexer();
docIndexer.doIndexing(context);
```

Reply:
[Offline Recap] - For this use case the WorkCoordinator has a slightly different use case than other parts of the system so lets run with this style.  Maybe we need to revisit in the future, but there will always be exceptions.  Marking this resolved.

Reply:
The underlying difference is that the WorkCoordination code is its own independent subsystem from the rest of document migration.  Work coordination could be used from a number of different calling contexts, so the contexts that are relevant for each method of coordination may have different context ancestries.  That's the reason that a context factory instead is sent to each method rather than to the constructor.  The methods themselves could take a context instance as an arg, but then it would be up to the caller to close/autoclose the resource.